plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemory.hs

```diff
@@ -94,13 +96,36 @@ type CostingInteger =
     SatInt
 #endif
 
-
 -- $(if finiteBitSize (0::SatInt) < 64 then [t|Integer|] else [t|SatInt|])
 
+{- | Divide one costing integer by another, "rounding upwards".  This is needed
+when we expose costs to the ledger, which will use different (smaller) units.
+Suppose one ledger unit is 1000 real cost units: then if a script costs 12345678
+real units we want that to be converted to 12346 ledger units, since 12345 (as
+given by `div`) would convert to 12345000 real units, which wouldn't be enough
+to run the script.  We want a <= (a `uppperIntegerQuotient` b) * b < a+b for all
+a and for all b>0, except that we may get '==' instead of '<' when we're using
+SatInt and a+b == MaxBound.  The behaviour when b <= 0 is unspecified.
+-}
+uppperIntegerQuotient :: CostingInteger -> CostingInteger -> CostingInteger
+a `uppperIntegerQuotient` b =
```

Comment:
You could implement this as `ceiling (toRational a) / (toRational b)` I guess?

Reply:
I still don't entirely trust the `SatInt` division operations, and going via `Rational` adds extra obscurity, so I'd prefer to keeps things fairly explicit.  I suppose we could bypass division in `SatInt` by explicitly convert the arguments to Integers and doing the calculations using those: it's not critical that this is fast, but it is critical that it's correct. Incidentally, if b < 0 then neither version your version or mine satisfies what I think the specification of this function should be (the problem being that then a+b < a, so the interval that the scaled-back-up result should lie in is empty).

Reply:
To be clear, I really want to be sure that this is doing the right thing (hence all the testing).  Introducing extra numeric types makes it less clear what's going on behind the scenes.  There's always a danger of denominators getting very large if you're dealing with rational numbers, and that might cause problems with our saturated integers: an apparently innocuous addition might introduce a `maxBound` somewhere that would later get reduced to something smaller without the overflow ever being visible to us.  That sounds dangerous when you're dealing with cost bounds.

Reply:
I don't think multiplying by a `Rational` is obscure!

I think my version using a single `scale` function is pretty clearly safe:
```
scale r a = 
  let
    ar :: Rational
    ar = toRational a
    m :: Rational
    m = r * ar
    c :: SatInt
    c = ceiling m
```
The specification of what this should do is clear, all the operations happen in `Rational`, which is well-behaved, and we just convert back to `SatInt` at the end inside `ceiling` (which converts to the final type at the end only). What's not to like?

Reply:
>  What's not to like?
I think it's just making things a bit too clever.  We're going through at least three different types just to multiply/divide two numbers.  I'd rather just keep things as explicit as possible.  It's only alike three lines of code anyway!

Reply:
Multiplying by a rational really doesn't seem that clever to me... Moreover, it's one line of code that I think is pretty obviously correct, means we don't have to worry about the zero denominator case, means we don't have to worry about the operations of `SatInt`, and probably doesn't need extensive testing (because it's not doing anything in `SatInt` really).